specreduce/wavelength_calibration.py

```diff
@@ -0,0 +1,230 @@
+from astropy.table import QTable, hstack
+import astropy.units as u
+from functools import cached_property
+from gwcs import wcs
+from gwcs import coordinate_frames as cf
+import numpy as np
+from specutils import Spectrum1D
+
+
+__all__ = ['WavelengthCalibration1D']
+
+
+def get_available_catalogs():
+    """
+    ToDo: Decide in what format to store calibration line catalogs (e.g., for lamps)
+          and write this function to determine the list of available catalog names.
+    """
+    return []
+
+
+def concatenate_catalogs():
+    """
+    ToDo: Code logic to combine the lines from multiple catalogs if needed
+    """
+    pass
+
+
+class WavelengthCalibration1D():
+
+    def __init__(self, input_spectrum, matched_line_list=None, line_pixels=None,
+                 line_wavelengths=None, catalog=None, model=Linear1D(), fitter=None):
+        """
+        input_spectrum: `~specutils.Spectrum1D`
+            A one-dimensional Spectrum1D calibration spectrum from an arc lamp or similar.
+        matched_line_list: `~astropy.table.QTable`, optional
+            An `~astropy.table.QTable` table with (minimally) columns named
+            "pixel_center" and "wavelength" with known corresponding line pixel centers
+            and wavelengths populated.
+        line_pixels: list, array, `~astropy.table.QTable`, optional
+            List or array of line pixel locations to anchor the wavelength solution fit.
+            Will be converted to an astropy table internally if a list or array was input.
+            Can also be input as an `~astropy.table.QTable` table with (minimally) a column
+            named "pixel_center".
+        line_wavelengths: `~astropy.units.Quantity`, `~astropy.table.QTable`, optional
+            `astropy.units.Quantity` array of line wavelength values corresponding to the
+            line pixels defined in ``line_list``. Does not have to be in the same order]
+            (the lists will be sorted) but does currently need to be the same length as
+            line_list. Can also be input as an `~astropy.table.QTable` with (minimally)
+            a "wavelength" column.
+        catalog: list, str, `~astropy.table.QTable`, optional
+            The name of a catalog of line wavelengths to load and use in automated and
+            template-matching line matching.
+        model: `~astropy.modeling.Model`
+            The model to fit for the wavelength solution. Defaults to a linear model.
+        fitter: `~astropy.modeling.fitting.Fitter`, optional
+            The fitter to use in optimizing the model fit. Defaults to
+            `~astropy.modeling.fitting.LinearLSQFitter` if the model to fit is linear
+            or `~astropy.modeling.fitting.LMLSQFitter` if the model to fit is non-linear.
+
+        Note that either ``matched_line_list`` or ``line_pixels`` must be specified,
+        and if ``matched_line_list`` is not input, at least one of ``line_wavelengths``
+        or ``catalog`` must be specified.
```

Comment:
this feels to me like it would be better as multiple classmethod, but we can revisit this as we iterate further on the API and various workflows as I recognize that might be a matter of taste:

* `WavelengthCalibration1D(input_spectrum, matched_line_list, model, fitter)` (`__init__`) 
* `WavelengthCalibration1D.from_lists(input_spectrum, line_pixels, line_wavelengths, model, fitter)`
* `WavelengthCalibration1D.from_catalog(input_spectrum, catalog, model, fitter)` (and perhaps `line_pixels`?)

This also would allow getting rid of a lot of code that just checks for validity of the combinations of all of these technically-optional but-sorta-required kwargs.

Reply:
yeah, this is still kinda messy. i think we should actually keep `WavelengthCalibration1D` as simple as possible: just `WavelengthCalibration1D(line_pixels, line_wavelengths, model, fitter)` where `line_pixels` and `line_wavelengths` are list-like of the same length. there are previous steps required to find/centroid lines in a calibration spectrum (`specutils` provides tools for this) and then match some of those lines to known wavelengths to construct `line_pixels` and `line_wavelengths` (manually or via to-be-implemented automated techniques). i don't think we should wrap all of these steps into one big class. the interfacing with things like catalogs of known lines like i implement in #165 will happen at the line-matching stage. 

anyway, it will be a lot easier to iterate on these ideas using realistic end-to-end examples once this work is merged with #165. the work may not be complete after these are merged, but nothing is broken so i don't think we need a separate `develop` branch`.